### ISSUE TITLE
no-extra-parens incorrectly flagging code that requires parens

### ISSUE THREAD

**Issue by michaelficarra** [eslint/eslint#4904]
Currently, `no-extra-parens` incorrectly does not report

``` js
(function(){}).a;
(function(){})[0];
(class{}).a;
(class{})[0];
```

This is due to [these lines](https://github.com/eslint/eslint/blob/a4e83124087c62d0dc4b64d55aa45721727c70e9/lib/rules/no-extra-parens.js#L420-L424), which I believe are a mistaken attempt to handle the ExpressionStatement restricted productions. Removing these lines will force the above program to be written as

``` js
(function(){}.a);
(function(){}[0]);
(class{}.a);
(class{}[0]);
```

which has no unnecessary parentheses. I'll see if I can whip up a PR.
##

**Comment by eslintbot** [comment]
@michaelficarra Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
Hmm, I'm not sure this should be changed. In all of these instances, parentheses are required somewhere. In other situations we flag, you can remove the parentheses and the syntax is valid. In these cases, its a syntax error to remove the parentheses unless you then include parentheses elsewhere. I see this as being a long of confusion for users as it appears this rule would be forcing a particular parentheses location rather than simply flagging unnecessary parentheses. So I'm :-1: on this change. 

@eslint/eslint-team anyone have thoughts?

**Comment by michaelficarra** [comment]
@nzakas It's not about enforcing a location, per se, but about precedence. `no-extra-parens` ensures you're not parenthesising a higher-precedence subexpression. In the above listed examples, these kinds of constructs were being allowed (but just for member expressions, strangely). If we wanted to take this path, then expressions like `(class{}) ? a : b;`, `(class{}), a;`, `(class{}) + a`, etc. should all be allowed at the top level, even though they all have parentheses that are unnecessary for precedence purposes. I strongly feel this is not the right way for this rule to handle ExpressionStatement restricted productions.

**Comment by nzakas** [comment]
>  ensures you're not parenthesising a higher-precedence subexpression

The rule as it exists does not do this. From the docs:

> This rule restricts the use of parentheses to only where they are necessary.

I think making a change so that this rule now considers precedence would be confusing. I don't want to tell users to remove parentheses if they will need to insert them somewhere else to avoid a syntax error. 

I think what you're asking for would actually be a separate rule that points out how to avoid unnecessary precedence changes.

**Comment by michaelficarra** [comment]
@nzakas I implemented the first version of the `no-extra-parens` rule, and I can tell you that it *almost exclusively* handles precedence. Every parenthesis that the rule will ever report is due to incorrect precedence. [Here's the code that finds precedence](https://github.com/eslint/eslint/blob/f5ebbae45e75f4ce0ca3c6d929842732cd2f77ef/lib/rules/no-extra-parens.js#L157-L236) and [here's the precedence tests](https://github.com/eslint/eslint/blob/f5ebbae45e75f4ce0ca3c6d929842732cd2f77ef/tests/lib/rules/no-extra-parens.js#L51-L113) (and [precedence failure tests](https://github.com/eslint/eslint/blob/f5ebbae45e75f4ce0ca3c6d929842732cd2f77ef/tests/lib/rules/no-extra-parens.js#L243-L305)).

> This rule restricts the use of parentheses to only where they are necessary **either because you're embedding a high-precedence expression in a low-precedence position or, infrequently, to avoid restricted grammar productions**.

As I said above, the current exception for MemberExpressions which start ExpressionStatements with a restricted production is a strange inconsistency in this rule.

**Comment by nzakas** [comment]
Fair enough, however, *almost exclusively* isn't *exclusively*. I can only see confusion coming from this change. Most users are not going to understand the nuance of this situation and will remove the parentheses only to be met with a syntax error. Then they'll probably add the parentheses right back where they were and begin the cycle again. 

As I said originally, parentheses in these cases are necessary and telling users that they aren't when what you actually mean is "please move these parentheses somewhere else" is just confusing.

While your proposal may create further consistency in what the rule flags, I don't think such consistency makes sense when it will create more confusion for users.

**Comment by michaelficarra** [comment]
Okay, then what do you recommend doing about this rule for the cases I mentioned above? It currently tells users to remove those parentheses, which will similarly give them a syntax error if they don't add them back elsewhere. Remember, the rule already has this behaviour for every other production in the grammar that starts with an expression.

We can make the rule smarter and warn them about a potential restricted production if users complain about this change. We can even have the rule automatically fix these problems if we want, and the potential for user confusion doesn't even come into play.

**Comment by nzakas** [comment]
I'm a bit confused. In the description, you stated that these patterns do not warn but should:

```js
(function(){}).a;
(function(){})[0];
(class{}).a;
(class{})[0];
```

I'm saying they should not warn since removing the parens results in a syntax error. These currently do not warn, so I'm unclear on what you're saying needs to be addressed. Are there other patterns you're talking about?

**Comment by michaelficarra** [comment]
@nzakas Those are all member expressions. You're looking at the wrong comment. I was referring to [this one](https://github.com/eslint/eslint/issues/4904#issuecomment-170377259), particularly the sentence

> If we wanted to take this path, then expressions like `(class{}) ? a : b;`, `(class{}), a;`, `(class{}) + a`, etc. should all be allowed at the top level, even though they all have parentheses that are unnecessary for precedence purposes.

Note that I *do not* recommend this. We should remove the special exception for MemberExpressions.

**Comment by nzakas** [comment]
Ah, I see what you mean, thanks for explaining. Context gets hard when you're jumping through tons of issues.

Again, I don't see a good argument to change the behavior of `MemberExpression`s. I don't see that as being expected by our users.

You make a good point about `(class{}) ? a: b;`. We probably do need to change those so they don't warn. I know that's not your recommended approach, and I hope you can see my point about user confusion in regard to this pattern as well.

**Comment by michaelficarra** [comment]
@nzakas You feel users will be confused even if we make the rule smarter as I suggested?

> If users complain about this change, we can make the rule smarter and also warn them about a potential restricted production in those cases. We can even have the rule automatically fix these problems if we want, and the potential for user confusion doesn't even come into play.

**Comment by nzakas** [comment]
@michaelficarra I'm sorry, I just disagree with your proposed changes. I've already explained several times my motivation and concerns. The only work I see to be done is what I mentioned in https://github.com/eslint/eslint/issues/4904#issuecomment-171462731. If you'd like to help with that, I'd appreciate it. If not, I understand. 

If you're interested in pursuing a new rule that can help with precedence warnings, please feel free to open a new issue for that.

**Comment by kaicataldo** [comment]
Where do we stand on this? This issue has been open for a long time and I think it's hard for a contributor to know what, if anything, should be done.

**Comment by michaelficarra** [comment]
I still strongly feel that the exception for MemberExpressions in this rule is an error. It allows sub-expressions of  MemberExpression that have higher precedence than a MemberExpression to be parenthesised. The goal of this rule is to warn about exactly those kinds of situations. And the only opposition was on the grounds of user confusion, but now that we have auto-fixing, that argument is completely irrelevant.

**Comment by nzakas** [comment]
The rule, as currently implemented and described the docs, is designed to point out parens that can safely be removed without changing how the code works. I still feel that any change where removing parens causes a syntax error is the wrong thing for end users. I also don't believe that autofix for this rule should ever have to insert parens, as I don't believe that is expected.

So I'm still strongly :-1: to this change.

**Comment by kaicataldo** [comment]
Seems like it might be time to close this since this has been open since January and consensus hasn't been reached.

**Comment by ilyavolodin** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
